python/pyarrow/tests/test_compute.py

```diff
@@ -1899,6 +1900,71 @@ def test_assume_timezone():
     result.equals(pa.array(expected))
 
 
+def _check_temporal_rounding(ts, values, unit):
+    unit_shorthand = {
+        "nanosecond": "ns",
+        "microsecond": "us",
+        "millisecond": "L",
+        "second": "s",
+        "minute": "min",
+        "hour": "H",
+        "day": "D"
+    }
+    ta = pa.array(ts)
+
+    for value in values:
+        frequency = str(value) + unit_shorthand[unit]
+        options = pc.RoundTemporalOptions(value, unit)
+
+        result = pc.ceil_temporal(ta, options=options).to_pandas()
+        expected = ts.dt.ceil(frequency)
+        if unit != "nanosecond":
+            np.testing.assert_array_equal(result, expected)
+
+        result = pc.floor_temporal(ta, options=options).to_pandas()
+        expected = ts.dt.floor(frequency)
+        np.testing.assert_array_equal(result, expected)
+
+        result = pc.round_temporal(ta, options=options).to_pandas()
+        expected = ts.dt.round(frequency)
+        np.testing.assert_array_equal(result, expected)
+
+
+# TODO: We should test on windows once ARROW-13168 is resolved.
+@pytest.mark.skipif(sys.platform == 'win32',
+                    reason="Timezone database is not available on Windows yet")
+@pytest.mark.parametrize('unit', ("nanosecond", "microsecond", "millisecond",
+                                  "second", "minute", "hour", "day"))
+@pytest.mark.pandas
+def test_round_temporal(unit):
+    values = (1, 2, 3, 4, 5, 6, 7, 10, 15, 24, 60, 250, 500, 750)
+    timestamps = [
+        # "1899-04-18 01:57:09.190202880",
+        # "1899-09-12 07:03:30.080325120",
+        # "1904-06-21 20:55:36.493869056",
```

Comment:
Are these still meant to be commented out?

Reply:
There are some odd changes in timezone offsets that are causing problems. I'll try changing `ceil` to match behaviour.

Reply:
This is the kind of error I'm seeing for these three timestamps:

expected:
```
Timestamp('1899-04-19 00:31:50+0553', tz='Asia/Kolkata')
Timestamp('1899-09-13 00:31:50+0553', tz='Asia/Kolkata')
Timestamp('1904-06-22 23:59:50+0521', tz='Asia/Kolkata')
````

actual:
```
Timestamp('1899-04-19 00:00:00+0553', tz='Asia/Kolkata'),
Timestamp('1899-09-13 00:00:00+0553', tz='Asia/Kolkata'),
Timestamp('1904-06-23 00:00:00+0521', tz='Asia/Kolkata')
```

Reply:
https://en.wikipedia.org/wiki/Calcutta_Time

Reply:
Oh boy. So, if I'm reading this right, it's rounding using the "modern" TZ definition instead of the "contemporary" one? Maybe we should just not test this case then, since this is down to whatever the timezone database for a system contains…? Or am I misunderstanding.

Reply:
I think it's that but I'm not 100% and I really want to be :))

Reply:
Sorry I mislabeled that (switch actual and expected). It appears my rounding in local time is not done correctly.

Reply:
I'm back to thinking Pandas only uses "modern" TZs while date.h uses a historical database.
I've added a [test for this in c++](https://github.com/apache/arrow/blob/e2b6d9f24d37e50bc62e23203e4f6ba6e9275d1f/cpp/src/arrow/compute/kernels/scalar_temporal_test.cc#L1912-L1945) and removed it from Python.